django/contrib/admin/templates/admin/edit_inline/tabular.html

```diff
@@ -50,7 +50,7 @@ <h2>{{ inline_admin_formset.opts.verbose_name_plural|capfirst }}</h2>
         {% for fieldset in inline_admin_form %}
           {% for line in fieldset %}
             {% for field in line %}
-              {% if field.is_readonly or not field.field.is_hidden %}
+              {% if not field.field.is_hidden %}
```

Comment:
The thought about whether or not a read_only field with a HiddenInput should be displayed is precisely the reason why I talked about "inconsistency" in my bug report. Because it's currently displayed in tabular inlines but not in stacked inlines (in the case of field being the single field of a line).

In my opinion, a field with a HiddenInput should not be displayed. So that's the reason for my change in tabular.html : the test looks more logical to me than the previous one (which I think was there just because it's the negation of the related test in the "original" <td> containing the hidden input for fields. In my original patch in the ticket, I talked about rendering an empty string for the read-only field when it is hidden, because I worked on version 2.1 (and the field dictionary representation was rendered in this <td>) ; but in the master branch, the new test introduced for the "original"<td> coped with this.

The test for the "original" <td> wants to render hidden input form fields (so adding _if not field.is_read_only_ is ok because a read_only field will not render a input form field).
The test for the "field-position" <td> is just for rendering the field value, whether or not it's a **real** form field or just the raw value (for read-only field), so it looks more logical to have _if not field.field.is_hidden_.

The PR I provided will also correct the behaviour for stacked inlines -> I will add a unittest for this.
However, there will be just a little inconsistency remaining:
* in stacked inlines, the read_only + hidden values are rendered in the html (but not visible because in a "hidden" div)
* in tabular inlines, the read_only + hidden values won't appear anyhere in the html.

I think we an leave with this inconsitency, but it would also be possible to adapt to have the same behaviour by
* rendering the <th> with "hidden" class for the column title of a hidden_field
* rendering the <td> with "hidden" class for the field cell
In such a case, may be it would make more sense to render the hidden form inputs in the related hidden cell rather than in the "original" <td>
Let me know if you think this would be better. (My first thought was to avoid as most as possible the modifications in the tabular.html template, as it would require projects redefining the template - such as django-nested-admin - to do the same modifications. But this is surely not a good reason not to do this modification if it mkes things clearer)

Reply:
Hi @antoinehumbert — thanks for the quick reply. 

If we can resolve the inconsistency, then super! But we need to be clear on the behaviour (and the behaviour change) both in the tests, and possibly the release notes. (If it's particularly obscure then we maybe don't need a release note, but if it's a behaviour someone might be **using** then a small _Breaking change_ note is appropriate.) 

We also need to make sure we didn't introduce a regression ref a9f5652113f0721a7066e359ae28d14692ea3c47. (We added a test for similar in 0aeb802cf054cb369646c871b53c93a83c1fa58a the other day.)

Reply:
Just pushed new unit tests for stacked inlines. (may be you can have a look at my "todo" in the added tests -> I do not know why assertInHTML does not work in thoses cases, despite the fact that searched HTML is effectively present !)

For the regression, do you have any suggestion for building such a test case ? As far as I can see, I cannot see how VariableDoesNotExist warning could be raised as the only accessed variables for a read-only field are **field.field.name** and **field.contents**, which are both defined on AdminReadonlyField.

Reply:
Hi @carltongibson , I finally changed the tabular.html template so that all fields values are rendered in HMTL, even for hidden fields. But, in the case of hidden fields, the corresponding \<th\> (containing field label) and \<td\> (containing field value) have the _hidden_ class, so that the column for hidden field is not visible by user.

This is more consistent with the behaviour of stacked inlines, which for hidden fields are always rendered in HTML, but not visible because contained in some div with the _hidden_ class.

... I also corrected the unittests for hidden fields in stacked inlines.

Reply:
I think i ll definitely add a unittest for the colspan case of non-field errors. My asumption on the fact that it should be equal to the total number of columns is incorrect. It must not take into account the hidden columns, so the new behaviour will give the same incorrect html as the former behaviour.

Reply:
Just pushed correction for correct computation of colspan for non-field error row in tabular inlines